applogic/src/api/message_cubit.rs

```diff
@@ -0,0 +1,195 @@
+// SPDX-FileCopyrightText: 2024 Phoenix R&D GmbH <hello@phnx.im>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::{pin::pin, sync::Arc};
+
+use flutter_rust_bridge::frb;
+use phnxcoreclient::{
+    clients::CoreUser,
+    store::{Store, StoreEntityId, StoreNotification, StoreOperation},
+    ConversationMessageId,
+};
+use tokio::sync::watch;
+use tokio_stream::{Stream, StreamExt};
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, error};
+
+use crate::{
+    util::{spawn_from_sync, Cubit, CubitCore},
+    StreamSink,
+};
+
+use super::{
+    types::{UiConversationMessage, UiConversationMessageId},
+    user::user_cubit::UserCubitBase,
+};
+
+#[frb(dart_metadata = ("freezed"))]
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct MessageState {
+    pub message: UiConversationMessage,
+}
+
+#[frb(opaque)]
+pub struct MessageCubitBase {
+    core: CubitCore<MessageState>,
+    store: CoreUser,
+}
+
+impl MessageCubitBase {
+    #[frb(sync)]
+    pub fn new(
+        user_cubit: &UserCubitBase,
+        message_id: UiConversationMessageId,
+        initial_state: MessageState,
+    ) -> Self {
+        let message_id = message_id.into();
+
+        let store = user_cubit.core_user.clone();
+        let store_notifications = store.subscribe();
+
+        let core = CubitCore::with_initial_state(initial_state);
+
+        MessageContext::new(store.clone(), core.state_tx().clone(), message_id)
+            .spawn(store_notifications, core.cancellation_token().clone());
+
+        Self { core, store }
+    }
+
+    // Cubit interface
+
+    #[frb(getter, sync)]
+    pub fn is_closed(&self) -> bool {
+        self.core.is_closed()
+    }
+
+    pub fn close(&mut self) {
+        self.core.close();
+    }
+
+    #[frb(getter, sync)]
+    pub fn state(&self) -> MessageState {
+        self.core.state()
+    }
+
+    pub async fn stream(&mut self, sink: StreamSink<MessageState>) {
+        self.core.stream(sink).await;
+    }
+
+    // Cubit methods
+
+    pub async fn mark_as_read(&self) -> anyhow::Result<()> {
+        let (conversation_id, timestamp) = {
+            let state = self.core.state_tx().borrow();
+            let message = &state.message;
+            if message.is_read {
```

Comment:
Is this the only reason we want to mark each message as re/unread?

So far, the logic has been that we used timestamps and did some debouncing/comparisons later.

Memorizing this for every message seems a bit wasteful and also more error-prone, but maybe I'm missing something

Reply:
Good point. I will move the `mark_as_read` functionality back to conversation. Most likely, I well have to bring the debouncer back.

Reply:
In commit https://github.com/phnx-im/infra/pull/282/commits/5e169e20442a87f9690f282e6eb15fb9e6a62fcc, moved `mark_as_read` to the `ConversationDetailsCubit`. The API is based on marking the conversation until a specific message id as read, instead as it was done before, until a specific timestamp. The latter is fragile because of possible truncation of fractional seconds and pointing in-between of messages.

Later we might consider to migrate the SQL schema s.t. `last_read` is a pointer to a message instead of being a timestamp.

Debouncing is now done via a simple watch channel.